src/uart.rs

```diff
@@ -0,0 +1,375 @@
+}
+
+#[cfg(riot_module_periph_uart_modecfg)]
+impl StopBits {
+    fn to_c(self) -> uart_stop_bits_t {
+        match self {
+            Self::One => uart_stop_bits_t_UART_STOP_BITS_1,
+            Self::Two => uart_stop_bits_t_UART_STOP_BITS_2,
+        }
+    }
+}
+
+/// This struct contains the `UART` device and handles all operation regarding it
+///
+/// [UART implementation]: https://doc.riot-os.org/group__drivers__periph__uart.html
+#[derive(Debug)]
+pub struct UartDevice {
+    dev: uart_t,
+}
+
+impl UartDevice {
+    /// Unsafety: To use this safely, the caller must ensure that the returned Self is destructed before &'scope mut F becomes unavailable.
+    unsafe fn construct_uart<'scope, F>(
+        index: usize,
+        baud: u32,
+        user_callback: &'scope mut F,
+    ) -> Result<Self, UartDeviceError>
+    where
+        F: FnMut(u8) + Send + 'scope,
+    {
+        let dev = macro_UART_DEV(index as c_uint);
+        uart_init(
+            dev,
+            baud,
+            Some(Self::new_data_callback::<'scope, F>),
+            user_callback as *mut _ as *mut c_void,
+        )
+        .negative_to_error()
+        .map(|_| Self { dev })
+        .map_err(UartDeviceError::from_c)
+    }
+
+
+    /// Tries to initialize the given `UART`. Returns a Result with rather `Ok<RMain>` where `RMain` is the value returned by the scoped main function
+    /// or a `Err<UartDeviceStatus>` containing the error
+    ///
+    /// This is the scoped version of [`new()`] that can be used if you want to use short-lived callbacks, such as
+    /// closures or anything containing references. The UartDevice is deconfigured when the internal main function
+    /// terminates. A common pattern around this kind of scoped functions is that `main` contains the application's
+    /// main loop, and never terminates (in which case the clean-up code is eliminated during compilation).
+    /// # Arguments
+    ///
+    /// * `dev` - The index of the hardware device
+    /// * `baud` - The used baud rate
+    /// * `user_callback` The user defined callback that gets called from the os whenever new data is received from the `UART`
+    /// * `main` The mainloop that is executed inside the wrapper
+    ///
+    /// # Examples
+    /// ```
+    /// use riot_wrappers::uart::UartDevice;
+    /// let mut cb = |new_data| {
+    ///     //do something here with the received data
+    /// };
+    /// let mut scoped_main = |self_: &mut UartDevice| loop {
+    ///    self_.write(b"Hello from UART")
+    /// };
+    /// let mut uart = UartDevice::new_scoped(0, 115200, &mut cb, scoped_main)
+    ///    .unwrap_or_else(|e| panic!("Error initializing UART: {e:?}"));
+    /// ```
+    pub fn new_scoped<'scope, F, Main, RMain>(
+        index: usize,
+        baud: u32,
+        user_callback: &'scope mut F,
+        main: Main,
+    ) -> Result<RMain, UartDeviceError>
+    where
+        F: FnMut(u8) + Send + 'scope,
+        Main: FnOnce(&mut Self) -> RMain,
```

Comment:
Sorry, just found a soundness hole in this:

A user might do something like this:

```
new_scoped(…, |_char| (), |self_| {
    let racy = 0;
    let shortlived_callback = |_char| { *racy = char };
    new_scoped(…, shortlived_callback, |self2_| {
        core::mem::swap(self_, self2_);
    });
    // at drop time it's effectively self_ that is dropped
    do_something_with(&mut racy);
    // wait for  input
    // goes to self2_, calls shortlived_callback, writes into racy
});
```

The two classical workarounds that occur to me off my head are using Pin<Self> (which we don't fully need but resolves the issue because they can't be swapped), or making 'scope part of Self (where for _without_rx and _static that'd be 'static), and then Main is `for<'a> Main: FnOnce(&mut UartDevice<'a>)` so it becomes a brand, and this module will by construction ensure that there are never two UartDevice that can be swapped.

A scope was even in the type back in https://github.com/RIOT-OS/rust-riot-wrappers/pull/39#discussion_r1141487628, which from how I view it now was right, but we just didn't manage to figure out *why* it was. The new_scoped etc. types could still live on the `<'static>` version of this, which should make it unambiguous for both the user and the compiler which to pick.

Reply:
Hmm, which version would you recommend? I have to admit, I've reached the end of my knowledge here.

Reply:
I think the version currently on `chrysn-impl_uart_wrapper` would work. That the example still works unmodified even though I altered the lifetimes accordingly is a good sign.